jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java

```diff
@@ -447,11 +446,46 @@ public JibContainerBuilder runPluginExtensions(
     }
   }
 
-  @Nullable
-  private JibGradlePluginExtension findConfiguredExtension(
-      List<JibGradlePluginExtension> extensions, String extensionClass) {
-    Predicate<JibGradlePluginExtension> matchesClassName =
-        extension -> extension.getClass().getName().equals(extensionClass);
-    return extensions.stream().filter(matchesClassName).findFirst().orElse(null);
+  @SuppressWarnings({"unchecked"})
+  private <T> ContainerBuildPlan runPluginExtension(
+      Optional<Class<T>> extraConfigType,
+      JibGradlePluginExtension<?> extension,
+      ExtensionConfiguration config,
+      ContainerBuildPlan buildPlan)
+      throws JibPluginExtensionException {
+    T extraConfig = null;
+    if (extraConfigType.isPresent() && config.getExtraConfiguration().isPresent()) {
+      try {
+        Action<T> action = (Action<T>) config.getExtraConfiguration().get();
+        extraConfig = project.getObjects().newInstance(extraConfigType.get());
+        action.execute(extraConfig);
+      } catch (ClassCastException ex) {
+        throw ex; // TODO(chanseok): provide helpful and actionable message
+      }
+    }
+
+    return ((JibGradlePluginExtension<T>) extension)
+        .extendContainerBuildPlan(
+            buildPlan,
+            config.getProperties(),
+            Optional.ofNullable(extraConfig),
+            () -> project,
+            new GradleExtensionLogger(this::log));
+  }
+
+  private JibGradlePluginExtension<?> findConfiguredExtension(
+      List<JibGradlePluginExtension<?>> extensions, ExtensionConfiguration config)
+      throws JibPluginExtensionException {
+    Predicate<JibGradlePluginExtension<?>> matchesClassName =
+        extension -> extension.getClass().getName().equals(config.getExtensionClass());
+    Optional<JibGradlePluginExtension<?>> found =
```

Comment:
is it possible to find multiple or is `findFirst` just to speed things up?

Reply:
It's not possible (or at least it won't happen in practice) to match multiple extensions with the same class name. It means the user has added multiple extensions classes of the same name on the classpath.

Reply:
Ah yeah, just wondering if we should detect that and tell the user? or just ignore it.

Reply:
Alright, I will check if this is possible and see what to do if so.